test/jdk/java/net/httpclient/HttpClientAuthRetryLimitTest.java

```diff
@@ -0,0 +1,136 @@
+    private static final int DEFAULT_RETRY_LIMIT = 3;
+    private static final int RETRY_LIMIT = Integer.getInteger(
+            "jdk.httpclient.auth.retrylimit", DEFAULT_RETRY_LIMIT);
+
+    private static Stream<Object> args() {
+        return Stream.of(
+                HttpClient.Version.HTTP_1_1,
+                HttpClient.Version.HTTP_2)
+                .flatMap(version -> Stream
+                        .of(false, true)
+                        .map(secure -> Arguments.of(version, secure)));
+    }
+
+    private static HttpClient.Builder createClient(boolean secure) {
+        HttpClient.Builder builder = HttpClient.newBuilder();
+        if (secure) {
+            builder.sslContext(SSL_CONTEXT);
+        }
+        return builder;
+    }
+
+    @ParameterizedTest
+    @MethodSource("args")
+    void testDefaultSystemProperty(HttpClient.Version version, boolean secure) throws Exception {
+
+        AtomicInteger requestCount = new AtomicInteger(0);
+
+        try (HttpTestServer httpTestServer = ((secure)? HttpTestServer.create(
+                version, SSL_CONTEXT): HttpTestServer.create(version))) {
+            final String requestUriScheme = secure ? "https" : "http";
+            final String requestPath = "/" + this.getClass().getSimpleName() + "/";
+            final String uriString = "://" + httpTestServer.serverAuthority() + requestPath;
+            final URI requestUri = URI.create(requestUriScheme + uriString);
+
+            HttpTestHandler httpTestHandler = t -> {
+                t.getResponseHeaders()
+                        .addHeader("WWW-Authenticate", "Basic realm=\"Test\"");
+                t.sendResponseHeaders(401,0);
+            };
+
+            httpTestServer.addHandler(httpTestHandler, requestPath);
+            httpTestServer.start();
+            try (
+                HttpClient client = createClient(secure)
+                        .authenticator(new Authenticator() {
+                            @Override
+                            protected PasswordAuthentication getPasswordAuthentication() {
+                                requestCount.incrementAndGet();
+                                return new PasswordAuthentication("username", "password".toCharArray());
+                            }
+                        })
+                        .build()) {
+                HttpRequest request = HttpRequest.newBuilder().version(version)
+                        .GET()
+                        .uri(requestUri)
+                        .build();
+                IOException exception = assertThrows(IOException.class, () -> client.send(
+                        request, HttpResponse.BodyHandlers.discarding()));
+                assertEquals("too many authentication attempts. Limit: " + RETRY_LIMIT, exception.getMessage());
+                assertEquals(RETRY_LIMIT > 0 ? RETRY_LIMIT : 0,
+                        RETRY_LIMIT > 0 ? requestCount.get():requestCount.decrementAndGet());
```

Comment:
Could you split that in two statements for better readability? The double ternary operator makes it difficult to parse.

```
    if (RETRY_LIMIT > 0) {
        assertEquals(...);
    } else {
        assertEquals(...);
    }
```